### ISSUE TITLE
[no-magic-numbers] Exempt numbers used with void operators

### ISSUE THREAD

**Issue by mattacosta** [typescript-eslint/typescript-eslint#567]
Prior to ES5 the `undefined` variable could be changed to some value other than `undefined`. As a consequence, some developers prefer to use `void 0` instead of the variable. While usage of the [void operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void) is already covered by the [no-void](https://eslint.org/docs/rules/no-void) rule, the no-magic-number rule can still flag the literal value.

The rule should either specifically exempt `void 0`, or have an `ignoreVoidExpressions` option that would exempt any number used as an expression given to the void operator.

A temporary workaround is to use the `ignore` option on the number `0` instead, but that's a relatively common magic number that should be checked.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-magic-numbers": "error"
  }
}
```

```TS
class C {
  protected p: number | undefined = void 0;
  public foo(bar?: string) {
    if (bar === void 0) {
      throw new Error('Argument is undefined');
    }
  } 
}
```

**Expected Result**
No error.

**Actual Result**
`No magic number: 0.`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.2.0` |

**Comment by bradzacher** [comment]
Two questions

Why are you using `void 0` instead of just using `undefined` (or `null`)?
It saves 3 characters so it seems like it's a weird shorthand. I haven't ever seen anyone write it in their code (ofc transpilers use it because it saves 3 characters).

why should it specifically exempt the void operator?
To me it looks like you've passed a magic number in to the void operator. Technically you could have written `void ZERO` and gotten the same output.

**Comment by mattacosta** [comment]
**Why are you using `void 0` instead of just using `undefined` (or `null`)?**

It's still used. For example, until recently, there were 1,515 occurrences of `void 0` in [microsoft/vscode (commit)](https://github.com/microsoft/vscode/commit/ef2547d547204c9a2ff6bf6c7edc7ae871193484).

From a technical perspective, TypeScript can still target old versions of JS where it is possible to assign some other value to `undefined` so it could theoretically still be needed.

As for `null` that's a whole different type, and not a drop-in replacement for `undefined`.

**Why should it specifically exempt the void operator?**

> To me it looks like you've passed a magic number in to the void operator. Technically you could have written void ZERO and gotten the same output.

I could have used anything, even something like `void foo()`. All operands result in `undefined`. Therefore, neither `<magic-number>` nor `<non-magic-number>` convey any meaning, and it is impossible for the rule to prevent an error here because there is never an error here.

**Comment by bradzacher** [comment]
If you follow the convention of never assigning anything to `undefined`, then `undefined` is only ever emitted by the JS engine, and `null` is only ever emitted by user code.

Unless I'm mistaken `null` and `undefined` are functionally interchangeable, so as long as your code doesn't do any strict checks like `x === void 0` / `x === undefined` / `typeof x === 'undefined'`, then it shouldn't matter which one you use?

Just for the record - I'm not against adding this behind a flag, I'm just trying to understand why you'd use the void operator at all.

**Comment by mattacosta** [comment]
> Just for the record - I'm not against adding this behind a flag, I'm just trying to understand why you'd use the void operator at all.

Like I said, as of ES5 there's no technical reason to anymore. As for why people still use it, I would assume that it's because there is no technical reason not to either; so it's ultimately just a matter of preference.

**Comment by Pajn** [comment]
>Unless I'm mistaken null and undefined are functionally interchangeable

They are very different beasts, using null instead of undefined should never be a suggestion without context.

Just the first two examples that sprung to mind:

```
JSON.stringify({foo: null}) // '{"foo":null}'
JSON.stringify({foo: undefined}) // '{}'
```

```
function add(a, b = 1) {
  return a + b
}

add(1) // 2
add(1, undefined) // 2
add(1, null) // 1
```

**Comment by bradzacher** [comment]
I was speaking in regards to the suggestion I put forward in that comment:

> never assign undefined to anything, then undefined is only ever emitted by the JS engine, and null is only ever emitted by user code.

Then the two are functionally interchangeable as they both represent an "empty" value. Though they have different meanings if you follow the above guideline.

---

`JSON.stringify` is not really a "language feature" per sae, and its handling of undefined is an implementation detail of the function (which stems from the fact that the JSON spec doesn't include `undefined`).

A much better (and a much more explicit) way of doing that same thing is via the `delete` operator, or by omitting the property altogether.

Passing `undefined` and having it disappear through `JSON.stringify` is a good way to trip developers up.

---

Default parameter values is one of the exact reasons that you should never ever pass `undefined` yourself. It's a potential trap for developers and makes your code way less understandable because your intent is no longer clear.

Also the fact that `undefined` passed to an argument with a default value causes the engine to use the default value is an implementation detail of the language (which stems from the fact that they needed to design the default value feature such that it was able to be polyfilled).

```ts
function doSomething(a, b) { ... }

doSomething(1, null)
doSomething(1, undefined)

////////
// now we enhance add
////////

function doSomething(a, b = 1) { ... }

// refactoring this one is clear - do nothing at all.
// the dev very intentionally passed null, meaning they want an "empty" value as the second arg.
doSomething(1, null)

// how do you refactor this?
// Were they using undefined as an empty value in the same way you would use null?
// Or are they relying on the fact that in future if a default value is added, it'll get used automatically?
doSomething(1, undefined)
```

**Comment by golopot** [comment]
Since the no-magic-numbers is extended from the core rule of the same name, and that the issue is not specific to typescript, maybe the option should be proposed at eslint core?

[demo that the issue affects the core rule](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLW1hZ2ljLW51bWJlcnM6IDIgKi9cbnZvaWQgMDsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7Im5vLXVuZGVmIjoyLCJuby10aHJvdy1saXRlcmFsIjoyfSwiZW52Ijp7fX19)

**Comment by bradzacher** [comment]
that's a good point @golopot - this would be better suited to the core rule.

**Comment by bradzacher** [comment]
As per @golopot's suggestion - I'm going to close this request.
We won't be adding this to this repo, because it's not related to typescript syntax, and as such can be patched into the core eslint rule.

If you feel strongly about this, please file an issue or a PR over at https://github.com/eslint/eslint
